### PR TITLE
[issue #5] SessionFactoryBuilder and PersistenceConfiguration implementation

### LINKED ISSUE #5: SessionFactoryBuilder and PersistenceConfiguration implementation
SessionFactoryBuilder and PersistenceConfiguration implementation

- Connection Pool init
- DataSource init
- Property file reading

### PR DESCRIPTION
**Pull request by financeasist**
- implemented main logic for parse persistence.yml with different persistence units
- implemented main logic for creating  SessionFactory with Hikari datasource
- covered by tests

### CODE REVIEW

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public PersistenceConfiguration.PersistenceUnitConfiguration readPropertiesFile(
+      String propertyFile, String persistenceUnit) {
+    ClassLoader contextClassLoader = this.getClass().getClassLoader();
+
+    log.trace("Looking up for the file: {}", propertyFile);
```
Comment (olehhaliak) [inline src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java]: I think it makes sense to use log-level `debug` here

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    var propertiesFile = getPropertyFile(propertyFile, contextClassLoader);
+
+    var mapper = new ObjectMapper(new YAMLFactory());
+    PersistenceConfiguration configuration;
+    try {
+      configuration = mapper.readValue(propertiesFile, PersistenceConfiguration.class);
+      log.trace("resolved persistence properties from {} file", propertiesFile.getName());
```
Comment (olehhaliak) [inline src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java]: debug?

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      String persistenceUnit) {
+    log.trace("Looking up for the persistence unit: {}", persistenceUnit);
+    var persistenceUnitConfiguration =
+        configuration.persistenceUnits()
+            .stream()
+            .filter(unit -> persistenceUnit.equals(unit.name()))
+            .findAny()
+            .orElseThrow(() -> new PersistenceUnitNotFoundException(
+                format("Persistence unit %s was not found", persistenceUnit)));
+    log.trace("Found and resolved configuration for persistence unit {} ", persistenceUnit);
```
Comment (olehhaliak) [inline src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java]: debug?

File: pom.xml [hunk 4]
```diff
@@ -14,27 +14,35 @@
...
+        <dependency>
+            <groupId>org.reflections</groupId>
+            <artifactId>reflections</artifactId>
+            <version>${org.reflections.version}</version>
+        </dependency>
         <dependency>
             <groupId>com.zaxxer</groupId>
             <artifactId>HikariCP</artifactId>
-            <version>${hikari.version}</version>
+            <version>${hikari.verision}</version>
```
Comment (olehhaliak) [inline pom.xml]: typo 'verision'

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/PersistenceConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * Record holds and provides access for basic parameters that are mandatory for data source configuration.
+     */
+    public record DataSource(String jdbcUrl, String userName, String password, Integer poolSize,
+                             String driverClassName) {
+      public DataSource(String jdbcUrl, String userName, String password, Integer poolSize,
+                        String driverClassName) {
+        this.jdbcUrl =
+            requireNonNull(jdbcUrl, "Missed required property 'jdbcUrl' in configuration file");
```
Comment (olehhaliak) [inline src/main/java/com/bobocode/blyznytsia/bibernate/config/PersistenceConfiguration.java]: "Missed required property '%s' in configuration file" repeated 6 times. Perhaps, it makes sense to make a constant instead

File: src/test/java/com/bobocode/blyznytsia/bibernate/bootstrap/DefaultSessionFactoryBuilderTest.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class DefaultSessionFactoryBuilderTest {
+
+  @Test
+  @DisplayName("Create session factory with datasource from default file")
```
Comment (olehhaliak) [inline src/test/java/com/bobocode/blyznytsia/bibernate/bootstrap/DefaultSessionFactoryBuilderTest.java:11]: JFYI: you can use `@DisplayNameGeneration(CamelCaseNameGenerator.class)` to automatically generate human-readable test names out of method names.
But anyway I really liked that you spend time writing concise names for each test

File: src/test/resources/invalid-persistence.yml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+persistenceUnits:
+  - name: invalid-property
+    dataSource:
+      jdbcUrl: jdbc:postgresql://localhost:5432/postgres
+      userName: postgres
+      password: postgres
+      driverClassName: org.postgresql.Driver
+      invalidProp: 5
```
Comment (olehhaliak) [inline src/test/resources/invalid-persistence.yml]: please, add new line

File: src/test/resources/provided-persistence.yml [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+      userName: sa
+      password: password
+      driverClassName: org.h2.Driver
+
+  - name: postgres
+    dataSource:
+      jdbcUrl: jdbc:postgresql://localhost:5432/postgres
+      userName: postgres
+      password: postgres
+      driverClassName: org.postgresql.Driver
```
Comment (olehhaliak) [inline src/test/resources/provided-persistence.yml:14]: new line

**Review by olehhaliak — CHANGES_REQUESTED** [review]

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/PersistenceConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        this.userName =
+            requireNonNull(userName, "Missed required property 'userName' in configuration file");
+        this.password =
+            requireNonNull(password, "Missed required property 'password' in configuration file");
+        this.poolSize = nonNull(poolSize) ? poolSize : 10;
```
Comment (vitshch) [inline src/main/java/com/bobocode/blyznytsia/bibernate/config/PersistenceConfiguration.java]: Create constant for default poolSize

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    log.trace("Looking up for the file: {}", propertyFile);
+    var propertiesFile = getPropertyFile(propertyFile, contextClassLoader);
+
+    var mapper = new ObjectMapper(new YAMLFactory());
+    PersistenceConfiguration configuration;
+    try {
+      configuration = mapper.readValue(propertiesFile, PersistenceConfiguration.class);
+      log.trace("resolved persistence properties from {} file", propertiesFile.getName());
+    } catch (UnrecognizedPropertyException ex) {
+      throw new PersistencePropertyUnrecognizedException(ex.getMessage());
```
Comment (vitshch) [inline src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java:70]: Propagate also the original exception into the custom exception, not only message

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  private File getPropertyFile(String resourceName, ClassLoader contextClassLoader) {
+    var propertyResource = contextClassLoader.getResource(resourceName);
+    if (propertyResource == null) {
+      String errorMessage = format("Persistence file %s was not found", resourceName);
+      log.error(errorMessage);
+      throw new PersistenceFileNotFoundException(errorMessage);
+    }
+    log.info("found persistence configuration file {}", resourceName);
```
Comment (vitshch) [inline src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java]: Typo. `found` -> `Found`

**Review by vitshch — CHANGES_REQUESTED** [review]

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java (hunk 1 above)
Comment (financeasist) [reply src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java]: changed to debug

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java (hunk 2 above)
Comment (financeasist) [reply src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java]: changed to debug

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java (hunk 3 above)
Comment (financeasist) [reply src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java]: changed to debug

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java (hunk 11 above)
Comment (financeasist) [reply src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java]: fixed

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java (hunk 10 above)
Comment (financeasist) [reply src/main/java/com/bobocode/blyznytsia/bibernate/config/reader/YamlPropertyReader.java:70]: why? I've checked it and the message is descriptive enough and contains all the necessary information

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/PersistenceConfiguration.java (hunk 9 above)
Comment (financeasist) [reply src/main/java/com/bobocode/blyznytsia/bibernate/config/PersistenceConfiguration.java]: created

File: src/test/resources/provided-persistence.yml (hunk 8 above)
Comment (financeasist) [reply src/test/resources/provided-persistence.yml:14]: added

File: src/test/resources/invalid-persistence.yml (hunk 7 above)
Comment (financeasist) [reply src/test/resources/invalid-persistence.yml]: added

File: src/test/java/com/bobocode/blyznytsia/bibernate/bootstrap/DefaultSessionFactoryBuilderTest.java (hunk 6 above)
Comment (financeasist) [reply src/test/java/com/bobocode/blyznytsia/bibernate/bootstrap/DefaultSessionFactoryBuilderTest.java:11]: ok, thanks. But I prefer to have separate names for method itself and display name.

File: src/main/java/com/bobocode/blyznytsia/bibernate/config/PersistenceConfiguration.java (hunk 5 above)
Comment (financeasist) [reply src/main/java/com/bobocode/blyznytsia/bibernate/config/PersistenceConfiguration.java]: extracted to constant

File: pom.xml (hunk 4 above)
Comment (financeasist) [reply pom.xml]: fixed

**Review by vitshch — APPROVED** [review]

**Review by olehhaliak — APPROVED** [review]

**Pull request merged**